File: src/main/java/org/pyjjs/scheduler/core/api/impl/actors/common/behaviours/Behaviour.java
package org.pyjjs.scheduler.core.api.impl.actors.common.behaviours;

import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import com.typesafe.config.Config;
import org.pyjjs.scheduler.core.api.impl.actors.common.ActorState;
import org.pyjjs.scheduler.core.api.impl.actors.common.ActorStateInteraction;
import org.pyjjs.scheduler.core.common.SystemConfigKeys;
import org.pyjjs.scheduler.core.common.locale.LangResolver;
import org.pyjjs.scheduler.core.common.locale.LocaleNotFoundException;
import org.pyjjs.scheduler.core.common.locale.LocaleResolver;
import org.pyjjs.scheduler.core.api.impl.actors.common.messages.Message;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import scala.concurrent.duration.FiniteDuration;

import java.util.Collection;
import java.util.concurrent.TimeUnit;

public abstract class Behaviour<T extends ActorState, M extends Message> {
    public final Logger BEHAVIOUR_LOG = LoggerFactory.getLogger(getClass());

    private ActorStateInteraction<T> actorStateInteraction;

    private T actorState;
    private LocaleResolver localeResolver;

    protected Behaviour(){}

    protected abstract void perform(M message);

    protected void answer(M inMessage, Message answerMessage) {
        inMessage.getSender().tell(answerMessage, getActorRef());
    }

    protected void send(ActorRef receiver, Message message) {
        receiver.tell(message, getActorRef());
    }

    protected void sendToAll(Collection<ActorRef> receivers, Message message) {
        receivers.stream().forEach((receiver) -> send(receiver, message));
    }

    protected ActorRef getActorRef() {
        return getActorState().getActorRef();
    }

    protected String getActorLocalName() {
        return getActorLocalName(getActorRef());
    }

    protected String getActorLocalName(ActorRef actorRef) {
        return actorRef.path().name();
    }

    protected void saveActorState(T actorState) {
        actorStateInteraction.saveActorState(actorState);
    }

    public T getActorState() {
        return actorState;
    }

    public void setActorStateInteraction(ActorStateInteraction<T> actorStateInteraction) {
        this.actorStateInteraction = actorStateInteraction;
        this.actorState = actorStateInteraction.getActorState();
        try {
            LangResolver langResolver = getLangResolver();
            this.localeResolver = LocaleResolver.get(langResolver);
        } catch (LocaleNotFoundException e) {
            BEHAVIOUR_LOG.warn("Locale init problem", e);
        }
    }

    private LangResolver getLangResolver() {
        ActorSystem actorSystem = getActorState().getActorSystem();
        ActorSystem.Settings actorSystemSettings = actorSystem.settings();
        Config systemConfig = actorSystemSettings.config();
        return new LangResolver(systemConfig.getString(SystemConfigKeys.DEFAULT_LOCALE_KEY));
    }

    protected void printMessage(String messageKey, Object... properties) {
        String message = messageKey;
        if(localeResolver != null) {
            message = localeResolver.getString(messageKey, properties);
        }
        BEHAVIOUR_LOG.info(message, properties);
    }

    public void sendToParent(Message message) {
        getActorState().getActorContext().parent().tell(message, getActorRef());
    }

    public void sendToResources(Message message) {
        getActorState().getActorContext()
                .system()
                .actorSelection("user/resources/*")
                .tell(message, getActorRef());
    }

    public void sendToResourceSupervisor(Message message) {
        getActorState().getActorContext()
                .system()
                .actorSelection("user/resources")
                .tell(message, getActorRef());
    }

    public void sendToTaskSupervisor(Message message) {
        getActorState().getActorContext()
                .system()
                .actorSelection("user/tasks")
                .tell(message, getActorRef());
    }

    public void sendToSelf(Message message) {
        send(getActorRef(), message);
    }

    public void sendByDelay(Message message, long delay, TimeUnit timeUnit) {
        getActorState().getActorSystem()
                .scheduler()
                .scheduleOnce(
                        FiniteDuration.apply(delay, timeUnit),
                        ()-> sendToSelf(message),
                        getActorState().getActorContext().dispatcher());
    }

    public void sendToTasks(Message message) {
        getActorState().getActorContext()
                .system()
                .actorSelection("user/tasks/*")
                .tell(message, getActorRef());
    }

    protected abstract Class<M> processMessage();
}

File: src/main/kotlin/org/pyjjs/scheduler/core/api/impl/actors/common/behaviours/Behaviour.kt
package org.pyjjs.scheduler.core.api.impl.actors.common.behaviours

import akka.actor.ActorRef
import org.pyjjs.scheduler.core.api.impl.actors.common.ActorState
import org.pyjjs.scheduler.core.api.impl.actors.common.ActorStateInteraction
import org.pyjjs.scheduler.core.api.impl.actors.common.messages.Message
import org.pyjjs.scheduler.core.common.SystemConfigKeys
import org.pyjjs.scheduler.core.common.locale.LangResolver
import org.pyjjs.scheduler.core.common.locale.LocaleResolver
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import scala.concurrent.duration.FiniteDuration
import java.util.concurrent.TimeUnit
import kotlin.reflect.KClass

abstract class Behaviour<T : ActorState, M : Message> protected constructor() {
    val BEHAVIOUR_LOG: Logger = LoggerFactory.getLogger(javaClass)

    private lateinit var actorStateInteraction: ActorStateInteraction<T>

    lateinit var actorState: T
        private set
    private var localeResolver: LocaleResolver? = null

    fun invoke(message: Message) {
        perform(processMessage().java.cast(message))
    }
    abstract fun perform(message: M)

    protected fun <M: Message> answer(inMessage: M, answerMessage: Message) {
        inMessage.sender!!.tell(answerMessage, actorRef)
    }

    protected fun send(receiver: ActorRef, message: Message) {
        receiver.tell(message, actorRef)
    }

    protected fun sendToAll(receivers: Collection<ActorRef>, message: Message) {
        receivers.forEach { receiver -> send(receiver, message) }
    }

    protected val actorRef: ActorRef
        get() = actorState.actorRef

    protected val actorLocalName: String
        get() = getActorLocalName(actorRef)

    protected fun getActorLocalName(actorRef: ActorRef): String {
        return actorRef.path().name()
    }

    protected fun saveActorState(actorState: T) {
        actorStateInteraction.saveActorState(actorState)
    }

    fun setActorStateInteraction(actorStateInteraction: ActorStateInteraction<T>) {
        this.actorStateInteraction = actorStateInteraction
        this.actorState = actorStateInteraction.getActorState()
        this.localeResolver = LocaleResolver.get(langResolver)
    }

    private val langResolver: LangResolver
        get() {
            val actorSystem = actorState.actorSystem
            val actorSystemSettings = actorSystem.settings()
            val systemConfig = actorSystemSettings.config()
            return LangResolver(systemConfig.getString(SystemConfigKeys.DEFAULT_LOCALE_KEY))
        }

    protected fun logMessage(messageKey: String, vararg properties: Any) {
        val message = localeResolver?.getString(messageKey, *properties) ?: messageKey
        BEHAVIOUR_LOG.info(message, *properties)
    }

    fun sendToParent(message: Message) {
        actorState.actorContext.parent().tell(message, actorRef)
    }

    fun publishToEventStream(message: Message) {
        actorState.actorContext.system().eventStream().publish(message)
    }

    fun sendToResources(message: Message) {
        actorState.actorContext.system().actorSelection("user/resources/*").tell(message, actorRef)
    }

    fun sendToResourceSupervisor(message: Message) {
        actorState.actorContext.system().actorSelection("user/resources").tell(message, actorRef)
    }

    fun sendToTaskSupervisor(message: Message) {
        actorState.actorContext.system().actorSelection("user/tasks").tell(message, actorRef)
    }

    fun sendToSelf(message: Message) {
        send(actorRef, message)
    }

    fun sendByDelay(message: Message, delay: Long, timeUnit: TimeUnit) {
        actorState.actorSystem.scheduler().scheduleOnce(
                FiniteDuration.apply(delay, timeUnit),
                { sendToSelf(message) },
                actorState.actorContext.dispatcher())
    }

    fun sendToTasks(message: Message) {
        actorState.actorContext.system().actorSelection("user/tasks/*").tell(message, actorRef)
    }

    fun scheduleNotification(notification: Message, timeKey: String) {
        val schedulerConfig = actorState.actorSystem.settings().config()
        val notificationDelayInMillis = schedulerConfig.getLong(timeKey)
        sendByDelay(notification, notificationDelayInMillis, TimeUnit.MILLISECONDS)
    }

    abstract fun processMessage(): KClass<out M>
}
